### PR TITLE
Add function

### PR DESCRIPTION
**Pull request by Ridwanayinde**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by abeddost** [comment]
Hello @Ridwanayinde, I will take on the review process for this implementation.

**Comment by Ridwanayinde** [comment]
@abeddost Thank you. I will be awaiting your feedback.

File: solutions/tests/test_add.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+"""
+
+
+import unittest
+from solutions.add import add  # Import the function to test
+
+
+class TestAddFunction(unittest.TestCase):
+
```
Comment (abeddost) [inline solutions/tests/test_add.py:19]: Missing Class Docstrings

Comment (abeddost) [inline solutions/tests/test_add.py:19]: Missing Class Docstrings

File: solutions/tests/test_add.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def test_add_zero(self):
+        """Test adding zero to a number"""
+        self.assertEqual(add(4, 0), 4)
+
+    def test_non_numbers(self):
+        """it should raise an AssertionError for non numbers input"""
+        with self.assertRaises(AssertionError):
+            add("a", 5)  # Invalid input: string instead of a number
+
+
```
Comment (abeddost) [inline solutions/tests/test_add.py:41]: One thing I would suggest is adding a few more edge and defensive test cases to make sure the function handles all situations. Here’s a list of some cases that could be useful to include:

Test for adding very large values
Test for adding very small values
Test for adding with precision handling
Test for adding numbers resulting in overflow
Test for adding numbers represented as strings (e.g., "3" + "5")
Test for adding numbers mixed with None values
Test for adding non-numeric iterable inputs (e.g., lists or dictionaries)
Test for adding with boolean values as inputs

File: solutions/add.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+Module contents: 
+    -add: adds two numbers and returns the result.
+    
+Created on 28.12.2024
+@author : Ridwan Ayinde
+
+"""
+
+
+def add(a: float, b: float) -> float:
```
Comment (abeddost) [inline solutions/add.py:15]: Could you rename the parameters from a and b to number1 and number2 for better clarity?

File: solutions/add.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    Returns -> float: the sum of a and b
+
+    Raises:
+        AssertionError: if any argument is not a number
+
+    >>> add(4, 6)
+    10.0
+    >>> add(0, 0)
+    0.0
+
```
Comment (abeddost) [inline solutions/add.py:31]: Could you add at least one more test.

File: solutions/tests/test_add.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#! usr/bin/python3
```
Comment (abeddost) [inline solutions/tests/test_add.py:1]: I think the shebang line is incorrect. It has to be like this:
#!/usr/bin/env python3

**Review by abeddost — COMMENTED** [review]
Hi @Ridwanayinde, I’ve made some suggestions. Can you try applying them? They might fix the CI checks.

**Comment by Ridwanayinde** [comment]
Thank you @abeddost I have effected your comments.

File: solutions/add.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    -add: adds two numbers and returns the result.
+
+Created on 28.12.2024
+@author : Ridwan Ayinde
+
+"""
+
+
+def add(num1: float, num2: float) -> float:
+    """Add two numbers.
```
Comment (abeddost) [inline solutions/add.py:16]: Can you also add this into your sentence?

and returns their sum.

File: solutions/add.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    Parameters:
+        num1:float, the first number
```
Comment (abeddost) [inline solutions/add.py:19]: Could you please improve this for better clarity as well?

num1: (int, float) the first number to be added
num2: (int, float) the second number to be added

File: solutions/add.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        num2:float, the second number
+
+    Returns -> float: the sum of num1 and num2
+
+    Raises:
+        AssertionError: if any argument is not a number
```
Comment (abeddost) [inline solutions/add.py:25]: Could you also revise this to make it clearer?

AssertionError: If `num1` or `num2` is not an integer or float.

File: solutions/add.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    >>> add(0, 0)
+    0.0
+    >>> add(7, -3)
+    4.0
+    >>> add(-5, -3)
+    -8.0
+
+    """
+    # Validate inputs
+    assert isinstance(num1, (int, float)), "num1 must be a number"
```
Comment (abeddost) [inline solutions/add.py:38]: Could you also revise this to make it clearer?

num1 must be a int or float.

File: solutions/add.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    assert isinstance(num2, (int, float)), "num2 must be a number"
```
Comment (abeddost) [inline solutions/add.py:39]: Could you also revise this to make it clearer?

num2 must be a int or float.

**Review by abeddost — COMMENTED** [review]
Hey Ridwan! Excellent work! I’ve shared some suggestions to enhance clarity.

**Comment by Ridwanayinde** [comment]
Hi @abeddost I have effected your comments. Thanks

**Review by abeddost — APPROVED** [review]
Hi @Ridwanayinde, I’ve reviewed your changes, and everything looks perfect! Thank you for your hard work and dedication. Great job!

I am merging this branch into the main now.
